File: R/cyclical-GP/periodic.R
# periodic GP regression example

# See:
# Roberts, S., Osborne, M., Ebden, M., Reece, S., Gibson, N., and Aigrain, S. 2013. Gaussian processes for time-series modelling. Proc. R. Soc. A 371(1984): 20110550. doi:10.1098/rsta.2011.0550.
# Eq. 20

# Not able to get TMB model to work!

GP_periodic <- function(x, b0 = 0, sigma = 0.1,
  gp_sigma = 0.2, gp_theta = 1.5, gp_T = 12) {
  Sigma <- sapply(x, function(x1) {
    sapply(x, function(x2) {
      gp_part1 <- pi * abs((x2 - x1) / gp_T)
      gp_part2 <- sin(gp_part1) ^ 2
      gp_sigma * gp_sigma * exp((- 1 / (2 * gp_theta * gp_theta)) * gp_part2)
    })
  })
  eta <- MASS::mvrnorm(mu = rep(0, length(x)), Sigma = Sigma)
  y <- rnorm(length(x), mean = b0 + eta, sd = sigma)
  list(x = x, y = y, eta = eta, Sigma = Sigma)
}

library(TMB)
compile("periodic.cpp")
dyn.load(dynlib("periodic"))

set.seed(1018)
out <- GP_periodic(1:20, sigma = 0.1, gp_sigma = 0.5, gp_theta = 0.1, gp_T = 6)
plot(out$x, out$eta, type = "l", ylim = range(union(out$y, out$eta)))
points(out$x, out$y)
abline(v = seq(1, 36, 6))

data <- list(x = out$x, y = out$y, gp_T = 6, flag = 1, Sigma = out$Sigma)
parameters <- list(
  b0 = 0,
  log_gp_sigma = log(0.5), 
  log_gp_theta = log(0.9),
  log_sigma = log(0.1), 
  eta = rep(0, length(out$x))
)
obj <- TMB::MakeADFun(data, parameters, DLL="periodic",
  random = "eta")
# obj <- normalize(obj, flag="flag")
opt <- stats::nlminb(obj$par, obj$fn, obj$gr)
sdr <- TMB::sdreport(obj)
sdr

r <- obj$report()
print(r)

File: R/cyclical-GP/periodic.cpp
#include <TMB.hpp>

template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_VECTOR(x);
  DATA_VECTOR(y);
  PARAMETER(b0);
  PARAMETER(log_gp_sigma);
  PARAMETER(log_gp_theta);
  PARAMETER(log_sigma);
  PARAMETER_VECTOR(eta)
  // DATA_INTEGER(flag);
  // DATA_INTEGER(gp_T);
  
  int n = y.size(); 
  
  Type gp_sigma = exp(log_gp_sigma);
  Type gp_theta = exp(log_gp_theta);
  Type sigma = exp(log_sigma);
  REPORT(sigma);
  REPORT(gp_sigma);
  REPORT(gp_theta);

  
  matrix<Type> C(n, n);
  C.setZero();  
  
  Type dist = 0.0;
  
  for(int i=0; i<C.rows(); i++)
    for(int j=0; j<C.cols(); j++) {
      if (x[j] >= x[i])
        dist = x[j] - x[i];
      else
        dist = x[i] - x[j];
      // std::cout << dist << "\n";
      Type gp_part1 = Type(3.141592) * (dist / Type(6.0));
      Type gp_part2 = sin(gp_part1) * sin(gp_part1);
      C(i, j) = gp_sigma * gp_sigma * exp((- Type(1.0) / (Type(2.0) * gp_theta * gp_theta)) * gp_part2);
    }
    
  REPORT(C);
  Type jnll = density::MVNORM(C)(eta);
  // if (flag == 0) return jnll;
  jnll-= dnorm(y, vector<Type>(b0 + eta), sigma, true).sum();
  
  // jnll-= dnorm(gp_sigma, Type(0), Type(2));
  // jnll-= dnorm(gp_theta, Type(0), Type(3));
  // jnll-= dnorm(b0, Type(0), Type(10));
  // jnll-= dnorm(sigma, Type(0), Type(1));
  // jnll-= log_gp_sigma;
  // jnll-= log_gp_theta;
  // jnll-= log_sigma;
  
  return jnll;
}
